subprojects/core/src/test/groovy/org/gradle/initialization/DefaultExceptionAnalyserTest.groovy

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright 2010 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.initialization
+
+import org.gradle.api.GradleScriptException
+import org.gradle.api.tasks.TaskExecutionException
+import org.gradle.groovy.scripts.Script
+import org.gradle.groovy.scripts.ScriptSource
+import org.gradle.internal.event.ListenerManager
+import org.gradle.internal.exceptions.Contextual
+import org.gradle.internal.exceptions.LocationAwareException
+import org.gradle.internal.exceptions.MultiCauseException
+import spock.lang.Specification
+
+class DefaultExceptionAnalyserTest extends Specification {
+
+    private final ListenerManager listenerManager = Mock(ListenerManager.class)
+    private final StackTraceElement element = new StackTraceElement("class", "method", "filename", 7)
+    private final StackTraceElement callerElement = new StackTraceElement("class", "method", "filename", 11)
+    private final StackTraceElement otherElement = new StackTraceElement("class", "method", "otherfile", 11)
+    private final StackTraceElement elementWithNoSourceFile = new StackTraceElement("class", "method", null, 11)
+    private final StackTraceElement elementWithNoLineNumber = new StackTraceElement("class", "method", "filename", -1)
+    private final ScriptSource source = Mock(ScriptSource.class)
+
+    def setup() {
+        source.getFileName() >> "filename"
+        source.getDisplayName() >> "build file filename"
+    }
+
+    def 'wraps original exception when it is not a contextual exception'() {
+        given:
+        Throwable failure = new RuntimeException()
+        DefaultExceptionAnalyser analyser = analyser()
+
+        when:
+        Throwable transformed = analyser.transform(failure)
+
+        then:
+        transformed instanceof LocationAwareException
+
+        LocationAwareException gse = (LocationAwareException) transformed
```

Comment:
I guess you made casting and types explicit to get IDE support, but it's not necessary